io/plugins/eu.esdihumboldt.hale.io.geopackage/src/eu/esdihumboldt/hale/io/geopackage/GeopackageInstanceWriter.java

```diff
@@ -0,0 +1,570 @@
+			}
+
+			if (file.exists() && file.length() == 0L) {
+				// convenience for overwriting empty existing file
+				file.delete();
+			}
+			if (!file.exists()) {
+				GeoPackageManager.create(file);
+			}
+			// open existing file
+			geoPackage = GeoPackageManager.open(file, true);
+
+//			geoPackage.beginTransaction(); //XXX not sure how to use this -> results in a SQL exception later on (SQLITE_BUSY)
+
+			if (true /* isWriteUnordered() */) {
+				// write instances as they come in
+				writeInstances(geoPackage, instances, progress, reporter);
+			}
+			else {
+				// write instances based on type order needed for insert
+				// (to avoid violating constraints)
+				/*
+				 * Set<TypeDefinition> sortedSet = getSortedSchemas(
+				 * getTargetSchema().getMappingRelevantTypes());
+				 * 
+				 * for (TypeDefinition td : sortedSet) {
+				 * writeInstances(geoPackage, instances.select(new
+				 * TypeFilter(td)), progress, reporter); }
+				 */
+			}
+
+//			geoPackage.commit();
+//			geoPackage.endTransaction();
+
+			reporter.setSuccess(true);
+		} catch (Exception e) {
+			reporter.error(new IOMessageImpl(e.getLocalizedMessage(), e));
+			reporter.setSuccess(false);
+			reporter.setSummary("Saving instances to GeoPackage failed.");
+		} finally {
+			if (geoPackage != null) {
+				geoPackage.close();
+			}
+			progress.end();
+		}
+
+		return reporter;
+	}
+
+	/**
+	 * Write instances to the GeoPackage.
+	 * 
+	 * @param geoPackage the GeoPackage
+	 * @param instances the instances to write
+	 * @param progress the progress indicator
+	 * @param reporter the reporter
+	 */
+	protected void writeInstances(GeoPackage geoPackage, InstanceCollection instances,
+			ProgressIndicator progress, IOReporter reporter) {
+		try (ResourceIterator<Instance> it = instances.iterator()) {
+			while (it.hasNext() && !progress.isCanceled()) {
+				Instance instance = it.next();
+				TypeDefinition type = instance.getDefinition();
+
+				// TODO only specific types? (e.g. ignore some kinds of types?)
+				String tableName = type.getName().getLocalPart();
+
+				// determine table for type (and create if necessary)
+				GeopackageTableType tableType = createTableIfNecessary(geoPackage, tableName, type,
+						instance, reporter);
+				switch (tableType) {
+				case ATTRIBUTE:
+					AttributesDao attributes = geoPackage.getAttributesDao(tableName);
+					AttributesRow arow = attributes.newRow();
+
+					populateRow(arow, instance);
+
+					attributes.insert(arow);
+					break;
+				case FEATURE:
+					FeatureDao features = geoPackage.getFeatureDao(tableName);
+					FeatureRow frow = features.newRow();
+
+					populateRow(frow, instance);
+
+					// set geometry
+					String geometryColumn = features.getGeometryColumnName();
+					// XXX instead of using value traverse (GeometryFinder) for
+					// geometry?
+					Object geom = new InstanceAccessor(instance).findChildren(geometryColumn)
```

Comment:
Would this be limited to certain geometry models right now?

Reply:
Generally right now everything is limited to what the JTS WKB reader/writer can handle (as internally JTS geometries are used), which seems to be less than what the GeoPackage library can.

Reply:
Generally right now everything is limited to what the JTS WKB reader/writer can handle (as internally JTS geometries are used), which seems to be less than what the GeoPackage library can.

Reply:
Generally right now everything is limited to what the JTS WKB reader/writer can handle (as internally JTS geometries are used), which seems to be less than what the GeoPackage library can.

Reply:
Generally right now everything is limited to what the JTS WKB reader/writer can handle (as internally JTS geometries are used), which seems to be less than what the GeoPackage library can.